tests/optimization/bfgs.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+   Copyright (c) 2009-2016, Ryan H. Lewis
+   All rights reserved.
+
+   This file is part of Elemental and is under the BSD 2-Clause License, 
+   which can be found in the LICENSE file in the root directory, or at 
+   http://opensource.org/licenses/BSD-2-Clause
+*/
+#include <utility>
+#include <El.hpp>
+#include <El/optimization/bfgs.hpp>
+#define CATCH_CONFIG_RUNNER
+#include <catch.hpp>
+
+using namespace El;
+
+template< typename T>
+std::pair< DistMatrix<T>, T>
+SimpleQuadraticBFGSTest( const Int & N){
+  const std::function< T(const DistMatrix<T>&)>
+  quadratic_function = [&](const DistMatrix<T>& theta){
+        return Dot(theta,theta)/T(2);
+  };
+
+  const std::function< DistMatrix<T>(const DistMatrix<T>&, DistMatrix<T>&)>
+  gradient = [&](const DistMatrix<T>& theta, El::DistMatrix<T>& y){
+      y = theta;
+      return y;
+  };
+  DistMatrix<T> x0( N, 1);
+  Ones(x0, N, 1);
+  auto val = El::BFGS( x0, quadratic_function,  gradient);
+  return std::make_pair( x0, val);
+}
+/***
+ *  f(x) = .5*(x'Ax - x'b)
+ *  grad(f) = Ax - b
+ * @param A
+ * @param x
+ * @return
+ */
+template< typename T>
+std::pair< DistMatrix<T>, T>
+QuadraticFunction( const DistMatrix<T> & A, const DistMatrix<T>& b)
+{
+  const std::function< T(const DistMatrix<T>&)>
+  quadratic_function = [&](const DistMatrix<T>& theta)
+  {
+        DistMatrix<T> y( theta);
+        Gemv(El::NORMAL, T(1), A, theta, T(0), y);
+        return (Dot(y,theta) - Dot(b, theta))/T(2);
+  };
+
+  const std::function< DistMatrix<T>(const DistMatrix<T>&, DistMatrix<T>&)>
+  gradient = [&](const DistMatrix<T>& theta, El::DistMatrix<T>& y)
+  {
+      Gemv(El::NORMAL, T(1), A, theta, T(0), y);
+      Axpy(T(-1), b, y);
+      return y;
+  };
+  DistMatrix<T> x0( A.Height(), 1);
+  Gaussian( x0, A.Height(), 1);
+  auto val = El::BFGS( x0, quadratic_function,  gradient);
+  return std::make_pair( x0, val);
+}
+
+template< typename T>
+std::pair< DistMatrix<T>, T>
+QuadraticBFGSTest(Int N){
+ DistMatrix<T> A,B;
+ Laplacian(A, N);
+ auto nrm = FrobeniusNorm(A);
+ Gemm(El::NORMAL, El::ADJOINT, T(1), A, A, B);
+ A = B;
+ A *= T(1)/(nrm*nrm);
+ DistMatrix<T> b( N, 1);
+ Gaussian(b, N, 1);
+ b *= (T(1)/FrobeniusNorm(b));
+ return QuadraticFunction<T>(A, b);
+}
+
+template< typename T>
+std::pair< DistMatrix<T>, T>
+RosenbrockTest(){
+    const std::function< T(const DistMatrix<T>&)>
+          rosenbrock = [&](const DistMatrix<T>& theta)
+    {
+        auto x1 = theta.Get(0,0);
```

Comment:
If `theta` is distributed as a `DistMatrix<T,STAR,STAR>` then one could call `GetLocal` instead.